ruby/lib/cucumber/query/hook_by_test_step.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+require_relative 'errors'
+
+module Cucumber
+  module Query
+    class HookByTestStep
+      def initialize(config)
+        @hook_id_by_test_step_id = {}
+
+        config.on_event :test_step_created, &method(:on_test_step_created)
+        config.on_event :hook_test_step_created, &method(:on_hook_test_step_created)
+      end
+
+      def hook_id(test_step)
+        return @hook_id_by_test_step_id[test_step.id] if @hook_id_by_test_step_id.key?(test_step.id)
+
+        raise TestStepUnknownError, "No hook found for #{test_step.id} }. Known: #{@hook_id_by_test_step_id.keys}"
```

Comment:
The design of the query object assumes that if something can't be found either an empty list, empty optional or null is returned. There are no methods that throw.

This is also why all the methods are name `findX` or `findAllX` or findZByX`.